src/libraries/System.Private.Uri/src/System/PercentEncodingHelper.cs

```diff
@@ -0,0 +1,158 @@
+﻿using System.Buffers;
+using System.Diagnostics;
+using System.Text;
+
+namespace System
+{
+    internal static class PercentEncodingHelper
+    {
+        public static unsafe int UnescapePercentEncodedUTF8Sequence(char* input, int length, ref ValueStringBuilder dest, bool isQuery, bool iriParsing)
+        {
+            // As an optimization, this method should only be called after the first character is known to be a part of a non-ascii UTF8 sequence
+            Debug.Assert(length >= 3);
+            Debug.Assert(input[0] == '%');
+            Debug.Assert(UriHelper.EscapedAscii(input[1], input[2]) != Uri.c_DummyChar);
+            Debug.Assert(UriHelper.EscapedAscii(input[1], input[2]) >= 128);
+
+            uint fourByteBuffer = 0;
+            int bytesLeftInBuffer = 0;
+            var fourByteSpan = new ReadOnlySpan<byte>(&fourByteBuffer, 4);
```

Comment:
A uint is used here as a buffer instead of stackalloc[4], so that we can rotate the buffer efficiently whenever we don't consume 4 bytes at a time

Reply:
Because `fourByteBuffer` is marked _address-taken_ (due to wrapping a span around it), every single read access to `fourByteBuffer` will involve a stack spill. Even something as innocuous as `fourByteBuffer >>= 8` will turn into: `mov tmp, dword ptr [addr_of_fourByteBuffer]; shr tmp, 8; mov dword ptr [addr_of_fourByteBuffer], tmp;`.

If being able to perform this type of bit twiddling efficiently is critical to the performance of this routine, consider not declaring `fourByteSpan` here and instead doing the following:

```cs
DecodeRune:

uint tmp = fourByteBuffer; // make a copy so that the *copy* (not the original) is marked address-taken
if (Rune.DecodeFromUtf8(new ReadOnlySpan<byte>(&tmp, bytesLeftInBuffer), ...) == ...) { /* ... */ }
```

Reply:
Changed, this added an extra 5% in throughput